src/main/java/net/blay09/mods/tcinventoryscan/client/ClientProxy.java

```diff
@@ -71,105 +75,116 @@ public void init(FMLInitializationEvent event) {
-                            lastScannedSlot = mouseSlot;
-                            currentScan = null;
-                        }
-                    } else {
-                        currentScan = null;
-                        lastScannedSlot = mouseSlot;
-                    }
-                } else if (isHoveringPlayer && currentScan != null) {
-                    ticksHovered++;
-                    if (ScanManager.isValidScanTarget(entityPlayer, currentScan, "@")) {
-                        if (ticksHovered > SOUND_TICKS && ticksHovered % 2 == 0) {
-                            entityPlayer.worldObj.playSound(
-                                    entityPlayer.posX,
-                                    entityPlayer.posY,
-                                    entityPlayer.posZ,
-                                    "thaumcraft:cameraticks",
-                                    0.2F,
-                                    0.45F + entityPlayer.worldObj.rand.nextFloat() * 0.1F,
-                                    false);
-                        }
-                        if (ticksHovered >= SCAN_TICKS) {
-                            try {
-                                if (ScanManager.completeScan(entityPlayer, currentScan, "@")) {
-                                    NetworkHandler.instance.sendToServer(new MessageScanSelf());
-                                }
-                            } catch (StackOverflowError e) {
-                                // Can't do anything about Thaumcraft freaking out except for calming it down if it
-                                // does.
-                                // If Thaumcraft happens to get into a weird recipe loop, we just ignore that and assume
-                                // the item unscannable.
-                            }
-                            ticksHovered = 0;
-                            currentScan = null;
-                        }
-                    }
-                }
-            } else {
-                ticksHovered = 0;
-                currentScan = null;
-                lastScannedSlot = null;
-            }
+    private boolean notHoldingThaumometer(EntityPlayer player) {
+        return player == null || player.inventory.getItemStack() == null
+                || player.inventory.getItemStack().getItem() != thaumometer;
+    }
+
+    private void cancel() {
+        ticksHovered = 0;
+        currentScan = null;
+        isValidSlot = false;
+    }
+
+    @SubscribeEvent
+    public void clientTick(TickEvent.ClientTickEvent ignored) {
+        if (!TCInventoryScanning.isServerSideInstalled) return;
+        // Get minecraft and player objects
+        Minecraft mc = Minecraft.getMinecraft();
+        EntityPlayer player = mc.thePlayer;
+        if (notHoldingThaumometer(player)) {
+            cancel();
+            return;
+        }
+        switch ((isValidSlot ? 0 : 1) + ((hoveringSlot == lastHoveredSlot ? 0 : 10))) {
```

Comment:
this weird if construct can probably be refactored into
```java
if (hoveringSlot == lastHoveredSlot) { 
    if (isValidSlot)
        a();
    else
        b(); 
} else {
    c();
}
```
which is far less confusing

Reply:
> this weird switch construct can probably be refactored into
> 
> ```java
> if (hoveringSlot == lastHoveredSlot) { 
>     if (isValidSlot)
>         a();
>     else
>         b(); 
> } else {
>     c();
> }
> ```
> 
> which is far less confusing

I agree. Switch statement got refactored.